docs: add open edx event guide (#80)
openedx/openedx-events: docs/how-tos/creating-new-events.rst

How to create a new Open edX Event
==================================

The mechanisms implemented by the Open edX Events library are supported and maintained by the Open edX community.

> Therefore, we've put together a guide on how to add a new event the library so future contributions are effective.

Reviewer: ```suggestion
Therefore, we've put together a guide on how to add a new event to the library so future contributions are effective.
```

> 1. Propose the new event to the community

Reviewer: [thoughts] I'm fine with this landing as-is, and iterating as needed, but @ormsbee was thinking more proprietary events might still live in this repo, possibly with a company specific part of the domain. Don't let this be a blocker to landing this PR. This is just a possible discussion for the future.

Author: What about this for now:

1. Adding optional to the step
2. Adding at the end:
_This step might be unnecessary for events not facing the Open edX community._

Reviewer: 1. Maybe land this as-is for now?
2. I wonder if we should create an open github issue and point to it in this doc with something like:
```
.. note::
    There is an open discussion about whether organization scoped events would also exist in openedx-events,
    in which case this step would be optional. See <insert issue link here>.
```

Author: That sounds good for now! 

BTW, there's an open issue about private event schemas that might also be related to this conversation. I'll leave it here just in case: https://github.com/openedx/openedx-events/issues/184

-----------------------------------------

When creating a new event, you must justify its implementation. For example, you could create a post in Discuss,
send a message through slack or open a new issue in the library repository listing your use cases for it. Or even,
if you have time, you could accompany your proposal with the implementation of the event to illustrate its behavior.

2. Place your event in an architecture subdomain
-------------------------------------------------

As specified in the Architectural Decisions Record (ADR) events naming and versioning, the event definition needs an Open edX Architecture
Subdomain for:

- The name of the event: ``{Reverse DNS}.{Architecture Subdomain}.{Subject}.{Action}.{Major Version}``
- The package name where the definition will live, eg. ``learning/`` or ``content_authoring/``.

> For those reasons, after studying your new event purpose, you must place it in one of the subdomains supported by the project:

Reviewer: Thank you for adding all of these details about subdomains.

I wonder if we could:
1. Keep the table limited to the subdomains in use in this project, and
2. Maybe warn that new subdomains may require some discussion, because there does not yet exist and agreed upon set on subdomains.

Also see this possible adjusted text:
```suggestion
For those reasons, after studying your new event purpose, you must place it in one of the subdomains already in use, or introduce a new subdomain:
```

+----------------+----------------------------------------------------------------------------------------------------+
| Subdomain name | Description                                                                                        |
+================+====================================================================================================+

> | Authoring      | Allows educators to create, modify, discover, package, annotate (tag), and share learning content. |

Reviewer: 1. Should we call this "Content Authoring" for consistency?
2. Although I know what you mean, I wonder if we should drop "discover" from this description, just to not add confusion with a potential future "discovery" subdomain.

+----------------+----------------------------------------------------------------------------------------------------+
| Learning       | Allows learners to consume content and perform actions in a learning activity on the platform.     |
+----------------+----------------------------------------------------------------------------------------------------+
| Discovery      | Allows learners to find the right content at the right time to help achieve their learning goals.  |
+----------------+----------------------------------------------------------------------------------------------------+
| Programs       | Allows educators and learners to manage and engage in bundled packages (programs) of learning.     |

> +----------------+----------------------------------------------------------------------------------------------------+

Reviewer: I think these would be dropped if we simply want to document subdomains in use, and not necessarily those that may be used in the future. That way this doc doesn't have to become the place to debate future subdomains.

Refer to `edX DDD Bounded Contexts <https://openedx.atlassian.net/l/cp/vf8XjRiX>`_ confluence page for more documentation on domain-driven design in the Open edX project.

3. Create the data attributes for the event (OEP-49)
----------------------------------------------------

Events send `data attributes <https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0049-django-app-patterns.html#data-py>`_ when triggered. Therefore, when designing your new event definition you must
decide if an existent data class works for your use case or you must create a new one. If the answer is the latter, then try to answer:

- Which attributes of the object are the most relevant?
- Which type are they?

> - Is any of them optional/required?

Reviewer: ```suggestion
- Are any of them optional/required?
```

Author: ups! thanks again!
